### PR TITLE
Reformat messages so messages become simple `list[ModelRequest | ModelResponse]`

### PR DESCRIPTION
**Pull request by samuelcolvin**
Followup to #232.

TODO:
* [x] update docs to reflect new behaviour (especially we no longer have clever logic for system detection)
* [x] correct logfire messages
* [ ] fix streaming - the order of messages seems to be wrong

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 1]
```diff
@@ -122,30 +114,42 @@ class RetryPrompt:
...
         if isinstance(self.content, str):
             description = self.content
         else:
-            json_errors = ErrorDetailsTa.dump_json(self.content, exclude={'__all__': {'ctx'}}, indent=2)
+            json_errors = error_details_ta.dump_json(self.content, exclude={'__all__': {'ctx'}}, indent=2)
             description = f'{len(self.content)} validation errors: {json_errors.decode()}'
         return f'{description}\n\nFix the errors and try again.'
 
 
+UserMessagePart = Annotated[Union[SystemPrompt, UserPrompt, ToolReturn, RetryPrompt], pydantic.Discriminator('kind')]
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/messages.py]: This is somewhat inconsistent - we should probably use `SystemPromptPart`, `UserPromptPart`, etc to be consistent with the `TextPart` and `ToolCallPart` below

Comment (sydney-runkle) [reply pydantic_ai_slim/pydantic_ai/messages.py]: I'll note, I don't really like `SystemPromptPart`, it's kind of verbose, but I do think we should be consistent across `UserMessagePart` and `ModelMessagePart`

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 2]
```diff
@@ -20,11 +21,8 @@ class SystemPrompt:
     content: str
     """The content of the prompt."""
 
-    role: Literal['user'] = 'user'
-    """Message source identifier, this type is available on all messages as either 'user' or 'model'."""
-
-    message_kind: Literal['system-prompt'] = 'system-prompt'
-    """Message type identifier, this type is available on all messages as a discriminator."""
+    kind: Literal['system-prompt'] = 'system-prompt'
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/messages.py]: Could retain the note about "available on all message parts as a discriminator", same for the others in this file

File: pydantic_ai_slim/pydantic_ai/agent.py [hunk 3]
```diff
@@ -428,7 +429,8 @@ async def main():
                             else:
                                 # continue the conversation
                                 handle_span.set_attribute('tool_responses', response_messages)
-                                response_msgs = ' '.join(r.message_kind for r in response_messages)
+                                # TODO this is wrong, it should be a list of parts
+                                response_msgs = ' '.join(r.role for r in response_messages)
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/agent.py]: Still TODO?

Comment (samuelcolvin) [reply pydantic_ai_slim/pydantic_ai/agent.py]: yes.

Comment (samuelcolvin) [reply pydantic_ai_slim/pydantic_ai/agent.py]: done.

**Pull request merged**